refactor: introduce violations

mkniewallner opened the pull request:
**PR Checklist**

-   [x] A description of the changes is added to the description of this PR.
-   [ ] If there is a related issue, make sure it is linked to this PR.
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [ ] Documentation in `docs` is updated

**Description of changes**

Preliminary refactoring to ease the implementation of https://github.com/fpgmaas/deptry/issues/349. It implements the concept of violations to hold information related to all found violations. Later on, the location of the module will be added, so we can add it in the text/JSON report.

fpgmaas commented:
Nice solution approach! One question; I currently think the `Violation` class is slightly confusing. The `issue_type` parameter is not really the issue type, but rather it is the detector that is able to detect the violation. What would you think of an alternative solution approach, where `Violation` is the base class, and we have other classes like `ObsoleteDependencyViolation`, `MissingDependencyViolation`, etc, that inherit from that?

Then within the issue detectors, we can replace

```
obsolete_dependencies.append(Violation(self.__class__, dependency))
```

with

```
obsolete_dependencies.append(ObsoleteDependencyViolation(dependency))
```

Then we can also define custom `to_text()` methods (or their `__str__` method), so we can handle them differently more easily in the reporters?

Curious for your thoughts!

mkniewallner commented:
> @mkniewallner Nice solution approach! One question; I currently think the `Violation` class is slightly confusing. The `issue_type` parameter is not really the issue type, but rather it is the detector that is able to detect the violation. What would you think of an alternative solution approach, where `Violation` is the base class, and we have other classes like `ObsoleteDependencyViolation`, `MissingDependencyViolation`, etc, that inherit from that?
> 
> Then within the issue detectors, we can replace
> 
> ```
> obsolete_dependencies.append(Violation(self.__class__, dependency))
> ```
> 
> with
> 
> ```
> obsolete_dependencies.append(ObsoleteDependencyViolation(dependency))
> ```
> 
> Then we can also define custom `to_text()` methods (or their `__str__` method), so we can handle them differently more easily in the reporters?
> 
> Curious for your thoughts!

That looks better indeed, you're right about the fact that `issue_type` does not really represent what it should. I'll try this alternative and get back to you.

mkniewallner commented:
Rebased to resolve conflicts and added https://github.com/fpgmaas/deptry/pull/354/commits/bbf2373caac8c4517c741d09c75f41e434497396 following your suggestion to subclass `Violation` per issue type. Much better indeed, thanks!
